### DISCUSSION TITLE
when or what will trigger the 'updated()' method?

### DISCUSSION THREAD

**Discussion by TIMESTICKING** [HomeSpan/HomeSpan#489] [Q&A]
I'm using the 1.7 version. In reference.md it says 
`returns true if a HomeKit Controller has requested an update to the value of the Characteristic, otherwise false. `
but when I was controlling LightBulb, I push the button on the home app, and the method `updated()` returned `false` still.

I want to make a function that when I toggle the button 3 times quickly, some actions will be taken. I guess I need to add a new service if this doesn't work.

**Comment by HomeSpan** [comment]
The updated() method will always return true if HomeSpan has received a message from the Home App to update that Characteristic.  To check how this works you can type "L2" into the Serial monitor.  This will show you every message received from or sent to the Home App.  Updates to Characteristics are generally sent by the Home App in HTTP PUT requests. 

Pressing a button rapidly in the Home App generally will not work.  The Home App may not have yet processed the first request before it gets the second request and the net result is no message sent.

I have experimented with this before to control lights whose pre-programmed color or temperature can be changed by rapidly turning the light on and then off a fixed number of times.  I usually find waiting 1/2 second or more between toggling a home app button works, but faster than that often does not.

**Reply by TIMESTICKING** [reply]
well Im positive that `updated()` method always return 0.
```
#pragma once

#include <Arduino.h>
#include "HomeSpan.h"

struct DEV_SWITCH : Service::LightBulb
{
    int LedPin;
    int flip = 0;
    SpanCharacteristic *power;

    uint8_t sta = 0;


    DEV_SWITCH(int ledpin) : Service::LightBulb() {
        power = new Characteristic::On();

        LedPin = ledpin;
        pinMode(ledpin, OUTPUT);
        LOG0("LED init on Pin: ");LOG0(ledpin);
    }

    int power_NewVal(){
        return flip ^ power->getNewVal();
    }

    boolean update(){
        digitalWrite(LedPin, power_NewVal());

        return true;
    }

    void loop(){
         LOG1(power->updated());
        if (sta == 0 && power->updated()){
            sta = 1; //start
            LOG0("<switch status> first\n");
        }else if (sta == 1 && power->updated() && power->timeVal() <= 1000){
            sta = 2;
            LOG0("<switch status> second\n");
        }else if (sta == 2 && power->updated() && power->timeVal() <= 1000){
            sta = 0; //reset
            power->setVal(!power->getVal());
            flip != flip;
            LOG0("switch status!\n");
        }else{
            sta = 0; //reset
        }
    }
};
```
here is my code

**Reply by HomeSpan** [reply]
You can't use the updated() method in the loop() method. It doesn't mean anything in that context.

When HomeSpan receives from the Home App any updates messages for one or more Characteristics in the same Service, it groups all update messages together and calls the the boolean update() function.  But how did you know which Characteristics for the Service were updated?  That's what the update() method is for.  It's used only inside the boolean update() function.  After the function returns with true, HomeSpan and the Home App itself take that as an acknowledgement the update is complete and the current value of each Characteristic is replaced by whatever new value was sent in the update.  

By the time you get to the loop() function all updates have already been processed, so the update method will always return zero within loop().

**Reply by TIMESTICKING** [reply]
Wow! that is such enlightenment! Really appreciate everything you've done!
I guess I have to try the `getVal()` method and the state machine to achieve what I want. THX!!! :)
